src/hotspot/share/gc/g1/g1HeapRegion.cpp

```diff
@@ -611,42 +629,36 @@ class G1VerifyLiveAndRemSetClosure : public BasicOopIterateClosure {
     }
     oop obj = CompressedOops::decode_raw_not_null(heap_oop);
 
-    LiveChecker<T> live_check(this, _containing_obj, p, obj, _vo);
+    LiveChecker<T> live_check(_failures, _containing_obj, p, obj, _vo);
     if (live_check.failed()) {
       live_check.report_error();
       // There is no point in doing remset verification if the reference is bad.
       return;
     }
 
-    RemSetChecker<T> remset_check(this, _containing_obj, p, obj);
+    RemSetChecker<T> remset_check(_failures, _containing_obj, p, obj);
     if (remset_check.failed()) {
       remset_check.report_error();
     }
   }
 
 public:
-  G1VerifyLiveAndRemSetClosure(G1CollectedHeap* g1h, VerifyOption vo) :
-    _vo(vo),
-    _containing_obj(nullptr),
-    _num_failures(0) { }
-
-  void set_containing_obj(oop const obj) {
-    assert(!G1CollectedHeap::heap()->is_obj_dead_cond(obj, _vo), "Precondition");
-    _containing_obj = obj;
+  G1VerifyLiveAndRemSetClosure(oop containing_obj, VerifyOption vo, G1VerifyFailureCounter* failures)
+    : _vo(vo),
+      _containing_obj(containing_obj),
+      _failures(failures) {
+    assert(containing_obj != nullptr, "must be");
+    assert(!G1CollectedHeap::heap()->is_obj_dead_cond(containing_obj, _vo), "Precondition");
```

Comment:
In this new structure, these two asserts are not very useful any more, IMO.

Reply:
Good point. The `is_obj_dead_cond` has already been checked by the caller. I can see why someone would like to keep the null-check. I'll see what @tschatzl says, and will make changes after that.

Reply:
Feel free to drop both.